agent/metrics/metrics.go

```diff
@@ -0,0 +1,214 @@
+		err := fmt.Errorf("Failed to create metric handler: missing projectID")
+		return nil, err
+	}
+	if instanceID == "" {
+		err := fmt.Errorf("Failed to create metric handler: missing instanceID")
+		return nil, err
+	}
+	if instanceZone == "" {
+		err := fmt.Errorf("Failed to create metric handler: missing instanceZone")
+		return nil, err
+	}
+	if metricDomain == "" {
+		err := fmt.Errorf("Failed to create metric handler: missing metricDomain")
+		return nil, err
+	}
+
+	startTime = time.Now()
+	codeCount = make(map[string]int64)
+
+	return &MetricHandler{
+		projectID:	projectID,
+		instanceID:	instanceID,
+		instanceZone:	instanceZone,
+		metricDomain:	metricDomain,
+		ctx:		ctx,
+		client:		client,
+	}, nil
+}
+
+func (h *MetricHandler) WriteMetric(metricType string, statusCode int) error {
+	if h == nil {
+		return nil
+	}
+	if metricType == h.GetResponseCountMetricType() {
+		err := h.writeResponseCodeMetric(statusCode)
+		return err
+	}
+	return nil
+}
+
+func (h *MetricHandler) GetResponseCountMetricType() string {
+	if h == nil {
+		return ""
+	}
+	return fmt.Sprintf("%s/instance/proxy_agent/response_count", h.metricDomain)
+}
+
+// writeResponseCodeMetric will gather response codes and write to cloud monarch once sample period is over
+func (h *MetricHandler) writeResponseCodeMetric(statusCode int) error {
+	responseCode := fmt.Sprintf("%v", statusCode)
+
+	// Update response code count for the current sample period
+	if _, ok := codeCount[responseCode]; ok {
+		codeCount[responseCode] += 1
+	} else {
+		codeCount[responseCode] = 1
+	}
+
+	// Only write time series after sample period is over
+	if time.Since(startTime) < samplePeriod {
+		return nil
+	}
+
+	// Write metric
+	log.Printf("WriteMetric|attempting to write metrics at time: %v\n", time.Now())
+	for responseCode, count := range codeCount {
+		responseClass := fmt.Sprintf("%sXX", responseCode[0:1])
+		timeSeries := h.newTimeSeries(h.GetResponseCountMetricType(), responseCode, responseClass, newDataPoint(count))
+
+		log.Printf("got %v occurances of %s response code\n", count, responseCode)
+		log.Printf("%v\n\n", timeSeries)
+
+		if err := h.client.CreateTimeSeries(h.ctx, &monitoringpb.CreateTimeSeriesRequest{
+			Name: monitoring.MetricProjectPath(h.projectID),
+			TimeSeries: []*monitoringpb.TimeSeries{timeSeries},
+		}); err != nil {
+			log.Println("Failed to write time series data: ", err)
+			return err
+		}
+	}
+
+	// Clean up
+	startTime = time.Now()
+	codeCount = make(map[string]int64)
+
+	return nil
+}
+
+// newTimeSeries creates and returns a new time series
+func (h *MetricHandler) newTimeSeries(metricType, responseCode, responseClass string, dataPoint *monitoringpb.Point) *monitoringpb.TimeSeries {
+        return &monitoringpb.TimeSeries{
+		Metric: &metricpb.Metric{
+			Type: metricType,
+			Labels: map[string]string{
+				"response_code": responseCode,
+				"response_code_class": responseClass,
+			},
+		},
+		Resource: &monitoredrespb.MonitoredResource{
```

Comment:
This seems to be the only point that `instanceID` and `instanceZone` are used.

Instead of passing these in via command line flags, let's fetch them from the metadata server similar to how the `utils` package will conditionally fetch VM identities if it sees that it is running on GCE.

Reply:
According to Yuri, we will not use this approach to avoid introducing extra dependencies on the metadata server